<launch>
  <!-- size of map, change the size inflate x, y, z according to your application -->
  <arg name="map_size_x"/>
  <arg name="map_size_y"/>
  <arg name="map_size_z"/>
  <arg name="init_x"/>
  <arg name="init_y"/>
  <arg name="init_z"/>
  <arg name="target_x"/>
  <arg name="target_y"/>
  <arg name="target_z"/>

  <arg name="drone_id"/>

  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>
  <arg name="point1_x"/>
  <arg name="point1_y"/>
  <arg name="point1_z"/>
  <arg name="point2_x"/>
  <arg name="point2_y"/>
  <arg name="point2_z"/>
  <arg name="point3_x"/>
  <arg name="point3_y"/>
  <arg name="point3_z"/>


  <!-- topic of your odometry such as VIO or LIO -->
  <arg name="odom_topic"/>

  <!-- number of moving objects -->
  <arg name="obj_num" value="10" />

  <!-- main algorithm params -->
  <include file="$(find co_navi_pkg)/launch/high_speed/run_files/advanced_param.xml">

    <arg name="drone_id" value="$(arg drone_id)"/>

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="odometry_topic" value="$(arg odom_topic)"/>

    <arg name="obj_num_set" value="$(arg obj_num)" />

    <!-- camera pose: transform of camera frame in the world frame -->
    <!-- depth topic: depth image, 640x480 by default -->
    <!-- don't set cloud_topic if you already set these ones! -->
    <arg name="camera_pose_topic" value="pcl_render_node/camera_pose"/>
    <arg name="depth_topic" value="pcl_render_node/depth"/>

    <!-- topic of point cloud measurement, such as from LIDAR  -->
    <!-- don't set camera pose and depth, if you already set this one! -->
    <arg name="cloud_topic" value="pcl_render_node/cloud"/>

    <!-- intrinsic params of the depth camera -->
    <arg name="cx" value="321.04638671875"/>
    <arg name="cy" value="243.44969177246094"/>
    <arg name="fx" value="387.229248046875"/>
    <arg name="fy" value="387.229248046875"/>

    <!-- maximum velocity and acceleration the drone will reach -->
    <arg name="max_vel" value="60.0" />
    <arg name="max_acc" value="20.0" />
    <arg name="max_jerk" value="8.0" />
    <!--always set to 1.5 times grater than sensing horizen-->
    <arg name="planning_horizon" value="100.0" /> 

    <arg name="use_distinctive_trajs" value="true" />

    <!-- 1: use 2D Nav Goal to select goal  -->
    <!-- 2: use global waypoints below  -->
    <arg name="flight_type" value="2" />
    
    <!-- global waypoints -->
    <!-- It generates a piecewise min-snap traj passing all waypoints -->
    <arg name="point_num" value="$(arg point_num)" />

    <arg name="point0_x" value="$(arg point0_x)" />
    <arg name="point0_y" value="$(arg point0_y)" />
    <arg name="point0_z" value="$(arg point0_z)" />

    <arg name="point1_x" value="$(arg point1_x)" />
    <arg name="point1_y" value="$(arg point1_y)" />
    <arg name="point1_z" value="$(arg point1_z)" />

    <arg name="point2_x" value="$(arg point2_x)" />
    <arg name="point2_y" value="$(arg point2_y)" />
    <arg name="point2_z" value="$(arg point2_z)" />

    <arg name="point3_x" value="$(arg point3_x)" />
    <arg name="point3_y" value="$(arg point3_y)" />
    <arg name="point3_z" value="$(arg point3_z)" />

    
  </include>

  <!-- trajectory server -->
  <node pkg="ego_planner" name="drone_$(arg drone_id)_traj_server" type="traj_server" output="screen">
    <remap from="position_cmd" to="drone_$(arg drone_id)_planning/pos_cmd"/>
    <remap from="~planning/bspline" to="drone_$(arg drone_id)_planning/bspline"/>

    <param name="traj_server/time_forward" value="1.0" type="double"/>
  </node>

  <!-- use simulator -->
  <include file="$(find co_navi_pkg)/launch/high_speed/run_files/simulator.xml">

    <arg name="drone_id" value="$(arg drone_id)"/>

    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>

    <arg name="init_x_" value="$(arg init_x)"/>
    <arg name="init_y_" value="$(arg init_y)"/>
    <arg name="init_z_" value="$(arg init_z)"/>

    <arg name="odometry_topic" value="$(arg odom_topic)" />
  </include>

</launch>
